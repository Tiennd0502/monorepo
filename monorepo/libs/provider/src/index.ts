export { default as TamaguiProvider } from './TamaguiProvider';
export { default as QueryClientProvider } from './QueryClientProvider';
