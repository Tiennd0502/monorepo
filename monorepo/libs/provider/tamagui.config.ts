import { createTamagui, createTokens } from 'tamagui';
import { config } from '@tamagui/config/v3';
import { themes } from '@tamagui/themes';

import {
  colorsMapping,
  radius,
  size,
  space,
  zIndex,
  defaultFont,
  headingFont,
} from '@monorepo/ui';

const tokens = {
  radius,
  size,
  space,
  zIndex,
};

const tamaguiConfig = createTamagui({
  ...config,
  defaultFont: 'body',
  themes,
  fonts: {
    body: defaultFont,
    heading: headingFont,
  },
  tokens: createTokens({
    ...tokens,
    color: colorsMapping,
  }),
});

type AppConfig = typeof config;

declare module 'tamagui' {
  type TamaguiCustomConfig = AppConfig;
}

export default tamaguiConfig;
