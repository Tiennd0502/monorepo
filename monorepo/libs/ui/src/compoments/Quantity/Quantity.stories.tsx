import { Meta, StoryObj } from '@storybook/react';

import Quantity from '.';

const meta: Meta<typeof Quantity> = {
  title: 'Components/Quantity',
  component: Quantity,
};

export default meta;

type Story = StoryObj<typeof Quantity>;

export const Default: Story = {
  args: {
    onChangeValue: () => null,
  },
};
