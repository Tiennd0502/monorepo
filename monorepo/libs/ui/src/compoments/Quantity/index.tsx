import { memo, useCallback, useMemo, useState } from 'react';

import Text from '../Text';
import { AddIcon, MinusIcon } from '../icons';
import { XStack, YStack } from 'tamagui';

interface QuantityProps {
  defaultValue?: number;
  onChangeValue: (value: number) => void;
}

const Quantity = ({ defaultValue = 0, onChangeValue }: QuantityProps) => {
  const [number, setNumber] = useState<number>(defaultValue);

  const isDisabled = useMemo(() => number <= 1, [number]);

  const handleIncrement = useCallback(() => {
    const result = number + 1;

    setNumber(result);
    onChangeValue(result);
  }, [number, onChangeValue]);

  const handleDecrement = useCallback(() => {
    setNumber(number - 1);
    onChangeValue(number - 1);
  }, [number, onChangeValue]);

  return (
    <XStack gap="$3" alignItems="center">
      <YStack
        testID='increment'
        alignItems="center"
        justifyContent="center"
        width="$7.5"
        height="$7.5"
        backgroundColor="$backgroundIcon"
        borderRadius="$2"
        onPress={handleIncrement}
      >
        <AddIcon />
      </YStack>
      <Text size="large" color="$primary" fontWeight="600">
        {number <= 9 ? '0' : ''}
        {number}
      </Text>
      <YStack
        testID='decrement'
        alignItems="center"
        justifyContent="center"
        disabled={isDisabled}
        onPress={handleDecrement}
        width="$7.5"
        height="$7.5"
        backgroundColor={
          isDisabled ? '$backgroundIconDisabled' : '$backgroundIcon'
        }
        borderRadius="$2"
      >
        <MinusIcon />
      </YStack>
    </XStack>
  );
};

export default memo(Quantity);
