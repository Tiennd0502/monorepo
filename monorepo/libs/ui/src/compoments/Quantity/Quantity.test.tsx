import {fireEvent, render, waitFor} from '@testing-library/react-native';

import { TamaguiProvider } from '@monorepo/provider';
import Quantity from '.';

const onChangeValue = jest.fn();

const baseProps = { onChangeValue };

const setup = (moreProps = {}) => {
  const props = {
    ...baseProps,
    ...moreProps,
  };
  
  return render(
    <TamaguiProvider>
      <Quantity {...props} />
    </TamaguiProvider>
  );
};

describe('Quantity', () => {

  it('should render quantity component', () => {
    const container = setup();

    expect(container).toMatchSnapshot();
  });

  it('should render quantity component with quantity greater than 9', () => {
    const container = setup({defaultValue: 10});

    expect(container).toMatchSnapshot();
  });

  it('should render quantity component with full screen', () => {
    const container = setup();

    expect(container).toMatchSnapshot();
  });

  it('calls onChangeValue when increment icon value clicked', () => {
    const {getByTestId} = setup();

    waitFor(() => {
      const incrementIcon = getByTestId('increment');
      fireEvent.press(incrementIcon);

      expect(onChangeValue).toHaveBeenCalled();
    });
  });

  it('calls onChangeValue when decrement icon value clicked', () => {
    const {getByTestId} = setup({defaultValue: 3});

    waitFor(() => {
      const decrementIcon = getByTestId('decrement');
      fireEvent.press(decrementIcon);

      expect(onChangeValue).toHaveBeenCalled();
    });
  });

  it('calls onChangeValue when decrement icon value clicked', () => {
    const {getByTestId} = setup({defaultValue: 2});

    waitFor(() => {
      const decrementIcon = getByTestId('decrement');
      fireEvent.press(decrementIcon);

      expect(onChangeValue).toHaveBeenCalled();
    });
  });

  it('No action when decrement icon clicked with default value zero', () => {
    const {getByTestId} = setup({defaultValue: 1});
    waitFor(() => {
      const decrementIcon = getByTestId('decrement');

      expect(decrementIcon.props.disabled).toEqual(true);
    });
  });
});
