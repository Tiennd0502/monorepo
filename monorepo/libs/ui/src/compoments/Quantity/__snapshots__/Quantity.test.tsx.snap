// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantity should render quantity component 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#e0e0e0",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "column",
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
    testID="increment"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="25"
      bbWidth="24"
      fill="none"
      focusable={false}
      height="25"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 25,
            "width": 24,
          },
        ]
      }
      testID="add-icon"
      vbHeight={25}
      vbWidth={24}
      width="24"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M19 11.5h-6v-6a1 1 0 10-2 0v6H5a1 1 0 100 2h6v6a1 1 0 102 0v-6h6a1 1 0 100-2z"
          fill={
            {
              "payload": 4280558628,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
        <RNSVGMask
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          height="17"
          maskContentUnits={1}
          maskUnits={1}
          width="16"
          x="4"
          y="4"
        >
          <RNSVGPath
            d="M19 11.5h-6v-6a1 1 0 10-2 0v6H5a1 1 0 100 2h6v6a1 1 0 102 0v-6h6a1 1 0 100-2z"
            fill={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          />
        </RNSVGMask>
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <Text
    allowFontScaling={false}
    style={
      {
        "color": "#242424",
        "fontFamily": "NunitoSans_7pt-SemiBold",
        "fontSize": 18,
        "lineHeight": 20,
      }
    }
    suppressHighlighting={true}
  >
    0
    0
  </Text>
  <View
    disabled={true}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#999",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "column",
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
    testID="decrement"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="24"
      bbWidth="24"
      fill="none"
      focusable={false}
      height="24"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      testID="minus-icon"
      vbHeight={24}
      vbWidth={24}
      width="24"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGRect
          fill={
            {
              "payload": 4280558628,
              "type": 0,
            }
          }
          height="1.6"
          propList={
            [
              "fill",
            ]
          }
          rx="1"
          width="16"
          x="4"
          y="11"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;

exports[`Quantity should render quantity component with full screen 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#e0e0e0",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "column",
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
    testID="increment"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="25"
      bbWidth="24"
      fill="none"
      focusable={false}
      height="25"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 25,
            "width": 24,
          },
        ]
      }
      testID="add-icon"
      vbHeight={25}
      vbWidth={24}
      width="24"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M19 11.5h-6v-6a1 1 0 10-2 0v6H5a1 1 0 100 2h6v6a1 1 0 102 0v-6h6a1 1 0 100-2z"
          fill={
            {
              "payload": 4280558628,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
        <RNSVGMask
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          height="17"
          maskContentUnits={1}
          maskUnits={1}
          width="16"
          x="4"
          y="4"
        >
          <RNSVGPath
            d="M19 11.5h-6v-6a1 1 0 10-2 0v6H5a1 1 0 100 2h6v6a1 1 0 102 0v-6h6a1 1 0 100-2z"
            fill={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          />
        </RNSVGMask>
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <Text
    allowFontScaling={false}
    style={
      {
        "color": "#242424",
        "fontFamily": "NunitoSans_7pt-SemiBold",
        "fontSize": 18,
        "lineHeight": 20,
      }
    }
    suppressHighlighting={true}
  >
    0
    0
  </Text>
  <View
    disabled={true}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#999",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "column",
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
    testID="decrement"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="24"
      bbWidth="24"
      fill="none"
      focusable={false}
      height="24"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      testID="minus-icon"
      vbHeight={24}
      vbWidth={24}
      width="24"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGRect
          fill={
            {
              "payload": 4280558628,
              "type": 0,
            }
          }
          height="1.6"
          propList={
            [
              "fill",
            ]
          }
          rx="1"
          width="16"
          x="4"
          y="11"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;

exports[`Quantity should render quantity component with quantity greater than 9 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#e0e0e0",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "column",
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
    testID="increment"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="25"
      bbWidth="24"
      fill="none"
      focusable={false}
      height="25"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 25,
            "width": 24,
          },
        ]
      }
      testID="add-icon"
      vbHeight={25}
      vbWidth={24}
      width="24"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M19 11.5h-6v-6a1 1 0 10-2 0v6H5a1 1 0 100 2h6v6a1 1 0 102 0v-6h6a1 1 0 100-2z"
          fill={
            {
              "payload": 4280558628,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
        <RNSVGMask
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          height="17"
          maskContentUnits={1}
          maskUnits={1}
          width="16"
          x="4"
          y="4"
        >
          <RNSVGPath
            d="M19 11.5h-6v-6a1 1 0 10-2 0v6H5a1 1 0 100 2h6v6a1 1 0 102 0v-6h6a1 1 0 100-2z"
            fill={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          />
        </RNSVGMask>
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <Text
    allowFontScaling={false}
    style={
      {
        "color": "#242424",
        "fontFamily": "NunitoSans_7pt-SemiBold",
        "fontSize": 18,
        "lineHeight": 20,
      }
    }
    suppressHighlighting={true}
  >
    10
  </Text>
  <View
    disabled={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#e0e0e0",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "column",
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
    testID="decrement"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="24"
      bbWidth="24"
      fill="none"
      focusable={false}
      height="24"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      testID="minus-icon"
      vbHeight={24}
      vbWidth={24}
      width="24"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGRect
          fill={
            {
              "payload": 4280558628,
              "type": 0,
            }
          }
          height="1.6"
          propList={
            [
              "fill",
            ]
          }
          rx="1"
          width="16"
          x="4"
          y="11"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
